src/mds/Server.cc

```diff
@@ -878,12 +878,25 @@ void Server::find_idle_sessions()
     std::list<Session*> to_stale;
 
     for (auto session : *(sessions_p1->second)) {
+      client_t client = session->get_client();
+
       if (session->last_cap_renew >= cutoff) {
 	dout(20) << "laggiest active session is " << session->info.inst
 		 << " and sufficiently new (" << session->last_cap_renew << ")"
 		 << dendl;
 	break;
+      } else if (session->stale_mark_deferred and cutoff - session->
+		 last_cap_renew < mds->mdsmap->get_session_autoclose()) {
+	continue;
+      } else if (not (session->stale_mark_deferred or mds->locker->
+	    is_revoking_any_caps_from(client) or session->
+	    is_session_flushing_msgs(client.v))) {
+	  dout(20) << "deferring marking session " << client.v << " stale "
```

Comment:
I'd like to avoid using stale_mark_deferred.  just check

if (session->last_cap_renew >= autoclose_cutoff &&
!(mds->locker->is_revoking_any_caps_from(client) ||
 session->is_session_flushing_msgs(client)) {
continue;
}

besides, code style of your code is not good. don't break "A->B->C()" into multiple lines.